File: solvers/madrich/bindings/src/insert_job.cpp
#include "insert_job.h"

bool insert_greedy(Job job, std::vector <Route> routes, Problem *problem) {
    std::cout << "Greedy insert started\n";
    for (auto route :routes) {
        for (int i = 0; i < route.len() + 1; i++) {
            std::vector <Job> jobs;
            for (int j = 0; j < i; j++) {
                jobs.push_back(route.jobs[j]);
            }
            jobs.push_back(job);
            for (int j = i; j < route.jobs.size(); j++) {
                jobs.push_back(route.jobs[j]);
            }
            Route nr = route.new_route(jobs);
            std::optional <State> state = problem->get_state(&nr);
            if (!state) {
                continue;
            }
            route.jobs = jobs;
            std::cout << "Greedy insert positive";
            return true;
        }
    }
    std::cout << "Greedy insert negative\n";
    return false;
}

bool insert_best(Job job, std::vector<Route> routes, Problem* problem)
{
    std::cout << "Insert started\n";
    std::tuple<int, int> best_insert = {-1, -1};
    int best_time = -1;
	int i = 0;
    for (auto route : routes) {
        int old_tt = route.travel_time;
        for (int j = 0; j < route.len() + 1; j++) {
            std::vector <Job> jobs;
            for (int k = 0; k < j; k++) {
                jobs.push_back(route.jobs[k]);
            }
            jobs.push_back(job);
            for (int k = j; k < route.jobs.size(); k++) {
                jobs.push_back(route.jobs[k]);
            }
            Route nr = route.new_route(jobs);
            std::optional <State> state = problem->get_state(&nr);
            if (!state) {
                continue;
            }
            int new_time = state.value().travel_time - old_tt;
            if ((best_time == -1) || (new_time < best_time)) {
                best_time = new_time;
                best_insert = std::make_tuple(i, j);
            }
        }
        i++;
    }
    if (std::get<0>(best_insert) == -1) {
        std::cout << "Insert negative\n";
        return false;
    }
    i = std::get<0>(best_insert);
    int j = std::get<1>(best_insert);
    Route route = routes[i];
    int sz = route.jobs.size();
    route.jobs.clear();
    for (int k = 0; k < j; k++) {
        route.jobs.push_back(route.jobs[k]);
    }
    route.jobs.push_back(job);
    for (int k = j; k < sz; k++) {
        route.jobs.push_back(route.jobs[k]);
    }
    std::cout << "Insert positive\n";
    return true;
}    
    
    
    
    

File: solvers/madrich/problems/mdvrp_demo/operators/insert_job.py
import logging
from sys import maxsize
from typing import List

from madrich.problems.mdvrp_demo.models import Problem, Route, Job, Tour
from madrich.problems.mdvrp_demo.operators.utils import Block

"""
def insert_greedy(job: Job, routes: List[Route], problem: Problem) -> bool:
    logging.info('Greedy insert started')
    for route in routes:
        for i in range(len(route) + 1):
            jobs = route.jobs[:i] + [job] + route.jobs[i:]
            state = problem.get_state(route.new_route(jobs))
            if state is None:
                continue

            route.jobs = jobs
            logging.info('Greedy insert positive')
            return True
    logging.info('Greedy insert negative')
    return False
"""

def insert_best(block : Block, job: Job, routes: List[Route], problem: Problem) -> bool:
    logging.info('Insert started')
    best_insert = (-1, -1, -1)
    best_time = maxsize

    for i, route in enumerate(routes):

        if not block.check_route(route):
            logging.info('\nBLOCKED: insert_job')
            block.mark_route(route, False)
            continue
        status = False
        old_tt = route.state.travel_time
        for k, track in enumerate(route.tracks):
            # проверяем на перевес
            state_track = problem.get_state_track(track, route)
            size_v = len(state_track.value)
            for t in range(size_v):
                if state_track.value[t] + job.value[t] > route.courier.value[t]:
                    logging.info('\n track cant insert this job')
                    continue

            for j in range(len(track) + 1):
                old_jobs = track.jobs
                new_jobs = track.jobs[:j] + [job] + track.jobs[j:]
                track.jobs = new_jobs

                state = problem.get_state(route)
                if state is None:
                    track.jobs = old_jobs
                    continue
                new_time = state.travel_time - old_tt
                if new_time < best_time:
                    best_time = new_time
                    best_insert = (i, j, k)
                    status = True
                track.jobs = old_jobs
        block.mark_route(route, status)

    if best_insert[0] == -1:
        logging.info('Insert negative')
        return False

    i, j, k = best_insert
    route = routes[i]
    route.tracks[k].jobs = route.tracks[k].jobs[:j] + [job] + route.tracks[k].jobs[j:]
    logging.info('Insert positive')
    return True

def insert(tour: Tour, problem: Problem) -> bool:
    logging.info(f'\nUnassigned insert started, routes:{len(tour)}')
    block = Block(tour)

    result = False
    for storage in tour.storages:
        done = []
        unassigned_jobs = storage.unassigned_jobs
        for job in unassigned_jobs:
            if insert_best(block, job, tour.routes, problem):
                logging.info('Inserted')
                done.append(job)
                result = True

        for job in done:
            storage.unassigned_jobs.remove(job)
    return result